File: timing_table.py
# timing_table.py
import numpy as np
from scipy.interpolate import RegularGridInterpolator


def clamp(x, lo, hi):
    return max(lo, min(x, hi))


def find_axis_index(axis, x):
    """
    模擬 OpenSTA Axis::findAxisIndex
    回傳 i，使得 axis[i] <= x <= axis[i+1]
    並且 i 永遠在 [0, len(axis)-2]
    """
    if x <= axis[0]:
        return 0
    if x >= axis[-1]:
        return len(axis) - 2
    return np.searchsorted(axis, x) - 1


def get_value_from_table(
    values, input_slew_index, output_load_index, input_slew, output_load
):

    axis1 = np.array(input_slew_index[0])  # index_1
    axis2 = np.array(output_load_index[0])  # index_2
    table = np.array(values)

    size1 = len(axis1)
    size2 = len(axis2)

    # === size1 == 1 ===
    if size1 == 1:
        if size2 == 1:
            return float(table[0, 0])
        j = find_axis_index(axis2, output_load)
        x2l, x2u = axis2[j], axis2[j + 1]
        dx2 = (output_load - x2l) / (x2u - x2l)
        y00 = table[0, j]
        y01 = table[0, j + 1]
        return float((1 - dx2) * y00 + dx2 * y01)

    # === size2 == 1 ===
    if size2 == 1:
        i = find_axis_index(axis1, input_slew)
        x1l, x1u = axis1[i], axis1[i + 1]
        dx1 = (input_slew - x1l) / (x1u - x1l)
        y00 = table[i, 0]
        y10 = table[i + 1, 0]
        return float((1 - dx1) * y00 + dx1 * y10)

    # === bilinear ===
    i = find_axis_index(axis1, input_slew)
    j = find_axis_index(axis2, output_load)

    x1l, x1u = axis1[i], axis1[i + 1]
    x2l, x2u = axis2[j], axis2[j + 1]

    dx1 = (input_slew - x1l) / (x1u - x1l)
    dx2 = (output_load - x2l) / (x2u - x2l)

    y00 = table[i, j]
    y10 = table[i + 1, j]
    y01 = table[i, j + 1]
    y11 = table[i + 1, j + 1]

    value = (
        (1 - dx1) * (1 - dx2) * y00
        + dx1 * (1 - dx2) * y10
        + dx1 * dx2 * y11
        + (1 - dx1) * dx2 * y01
    )

    return float(value)


import numpy as np

import numpy as np


def get_1d_value_from_table(values, axis_index, target_value):

    # 假設 axis_index 的格式被包裝在一層 list/tuple 中
    axis = np.array(axis_index[0])

    # 加上 .flatten() 確保無論輸入是 [1,2,3] 還是 [[1,2,3]]，都會變成單純的 1D array
    table = np.array(values).flatten()

    size = len(axis)

    # === 當表格只有一個值時 (size == 1) ===
    if size == 1:
        return float(table[0])

    # === 線性插值 (Linear Interpolation) ===
    i = find_axis_index(axis, target_value)

    # 取得區間的上下界
    xl, xu = axis[i], axis[i + 1]

    # 計算目標值在該區間的比例 (0.0 ~ 1.0)
    dx = (target_value - xl) / (xu - xl)

    # 取得對應的 Y 值 (此時 table 已經是 1D，可以直接用 [i] 讀取)
    y0 = table[i]
    y1 = table[i + 1]

    # 透過距離比例進行加權計算
    value = (1 - dx) * y0 + dx * y1

    return float(value)

File: cppver/src/graph/timing_table.cpp
#include "easysta/graph/timing_table.hpp"

#include <algorithm>

namespace easysta::graph {

int find_axis_index(const std::vector<double>& axis, double x) {
    if (x <= axis.front()) return 0;
    if (x >= axis.back()) return static_cast<int>(axis.size()) - 2;
    auto it = std::lower_bound(axis.begin(), axis.end(), x);
    return static_cast<int>(it - axis.begin()) - 1;
}

double get_value_from_table(const std::vector<std::vector<double>>& values,
                             const std::vector<double>& input_slew_index,
                             const std::vector<double>& output_load_index,
                             double input_slew, double output_load) {
    const auto& axis1 = input_slew_index;
    const auto& axis2 = output_load_index;
    const auto& table = values;

    size_t size1 = axis1.size();
    size_t size2 = axis2.size();

    if (size1 == 1) {
        if (size2 == 1) {
            return table[0][0];
        }
        int j = find_axis_index(axis2, output_load);
        double x2l = axis2[j], x2u = axis2[j + 1];
        double dx2 = (output_load - x2l) / (x2u - x2l);
        double y00 = table[0][j];
        double y01 = table[0][j + 1];
        return (1 - dx2) * y00 + dx2 * y01;
    }

    if (size2 == 1) {
        int i = find_axis_index(axis1, input_slew);
        double x1l = axis1[i], x1u = axis1[i + 1];
        double dx1 = (input_slew - x1l) / (x1u - x1l);
        double y00 = table[i][0];
        double y10 = table[i + 1][0];
        return (1 - dx1) * y00 + dx1 * y10;
    }

    int i = find_axis_index(axis1, input_slew);
    int j = find_axis_index(axis2, output_load);

    double x1l = axis1[i], x1u = axis1[i + 1];
    double x2l = axis2[j], x2u = axis2[j + 1];

    double dx1 = (input_slew - x1l) / (x1u - x1l);
    double dx2 = (output_load - x2l) / (x2u - x2l);

    double y00 = table[i][j];
    double y10 = table[i + 1][j];
    double y01 = table[i][j + 1];
    double y11 = table[i + 1][j + 1];

    return (1 - dx1) * (1 - dx2) * y00
         + dx1 * (1 - dx2) * y10
         + dx1 * dx2 * y11
         + (1 - dx1) * dx2 * y01;
}

double get_1d_value_from_table(const std::vector<std::vector<double>>& values,
                                const std::vector<double>& axis_index,
                                double target_value) {
    const auto& axis = axis_index;

    // Flatten values (liberty 1D tables may be stored as a single row).
    std::vector<double> table;
    for (const auto& row : values) {
        table.insert(table.end(), row.begin(), row.end());
    }

    size_t size = axis.size();
    if (size == 1) {
        return table[0];
    }

    int i = find_axis_index(axis, target_value);
    double xl = axis[i], xu = axis[i + 1];
    double dx = (target_value - xl) / (xu - xl);

    double y0 = table[i];
    double y1 = table[i + 1];

    return (1 - dx) * y0 + dx * y1;
}

} // namespace easysta::graph
